src/records.rs

```diff
@@ -0,0 +1,303 @@
+            (ro.asset_def.code, ro.pub_key.address(), ro.freeze_flag),
+            (record.ro.amount.into(), record.uid),
+        ));
+        self.nullifier_records
+            .insert((record.nullifier, record.uid));
+        let mut editor = RecordEditor::new(self, record);
+        editor.save()?;
+        Ok(editor)
+    }
+
+    /// Rebuild the indices for this store
+    fn reload(&mut self) {
+        self.asset_records = Persistable::new();
+        self.nullifier_records = Persistable::new();
+        for record in self.store.iter() {
+            let ro = &record.ro;
+            self.asset_records.insert((
+                (
+                    ro.asset_def.code,
+                    ro.pub_key.address().clone(),
+                    ro.freeze_flag,
+                ),
+                (record.ro.amount.into(), record.uid),
+            ));
+            self.nullifier_records
+                .insert((record.nullifier, record.uid));
+        }
+    }
+
+    fn store<L: Ledger>(&mut self, uid: u64, record: &Record) -> Result<(), KeystoreError<L>> {
+        self.store.store(&uid, record)?;
+        self.reload();
+        Ok(())
+    }
+
+    /// Returns an Iterator to all the Records in the store
+    pub fn iter(&self) -> impl Iterator<Item = Record> + '_ {
+        self.store.iter().cloned()
+    }
+
+    /// Get a record from a uid from the store
+    pub fn get<L: Ledger>(&self, uid: u64) -> Result<Record, KeystoreError<L>> {
+        Ok(self.store.load(&uid)?)
+    }
+
+    /// get a Record Editor by uid from storage
+    pub fn get_mut<L: Ledger>(&mut self, uid: u64) -> Result<RecordEditor, KeystoreError<L>> {
+        let record = self.store.load(&uid)?;
+        Ok(RecordEditor::new(self, record))
+    }
+
+    /// Get records associated with an asset and account which are either frozen or unfrozen
+    /// Useful for finding records for transaction inputs
+    pub fn with_asset<L: Ledger>(
+        &self,
+        asset: &AssetCode,
+        owner: &UserAddress,
+        frozen: FreezeFlag,
+    ) -> Result<impl Iterator<Item = Record> + '_, KeystoreError<L>> {
+        let unspent_records = self
+            .asset_records
+            .index()
+            .get(&(*asset, owner.clone(), frozen))
+            .ok_or(KeyValueStoreError::KeyNotFound)?;
+        Ok(unspent_records
+            .iter()
+            .rev()
+            .map(move |(_, uid)| self.get::<L>(*uid).unwrap()))
+    }
+
+    /// Get one record with the exact amount or return None.  
+    pub fn with_asset_and_amount<L: Ledger>(
+        &self,
+        asset: &AssetCode,
+        owner: &UserAddress,
+        frozen: FreezeFlag,
+        amount: U256,
+        now: u64,
+    ) -> Result<Option<Record>, KeystoreError<L>> {
+        if let Some(unspent_records) =
+            self.asset_records
+                .index()
+                .get(&(*asset, owner.clone(), frozen))
+        {
+            if let Ok(amount) = amount.try_into() {
```

Comment:
Since this function explicitly deals with single records only, would it make more sense interface-wise for it to take a `RecordAmount`, and move the `try_into` into the caller?